merge_streams: refactor and reorder.

asah opened the pull request:
prep commit as per this code review:
https://github.com/zulip/zulip/pull/21821#discussion_r887398851

alexmv reviewed (commented):
I think I agree with Tim's comment in https://github.com/zulip/zulip/pull/21821#discussion_r887398851 that we should in general reorder this to be in the "safest" order possible.  Which I think is:
1. Subscribe people to the new stream
2. Move the messages
3. Unsubscribe everyone from the old stream
4. Deactivate the stream

I can see an argument for swapping 1/2 (not subscribing people to a random stream, if the message move fails), but I think the above order preserves that the messages never go missing for anyone, which seems like the right constraint.

I'm not convinced that catching the `JsonableError` and re-raising just its `str` as a `CommandError` is helping at all -- it strictly loses information which may be useful in understanding the error.

asah commented:
DONE. Note: given the number of changes, I upgraded merge_streams to a first class method of streams.py, added a unittest and converted merge_streams.py into a thin wrapper.

alexmv commented on zerver/tests/test_realm.py:
```diff
@@ -445,6 +445,31 @@ def test_get_default_notifications_stream(self) -> None:
 
+    def test_merge_streams(self) -> None:
+        realm = get_realm("zulip")
+        Denmark = get_stream("Denmark", realm)
```
```suggestion
        denmark = get_stream("Denmark", realm)
```

alexmv commented on zerver/actions/streams.py:
```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
+        message_ids_to_clear = message_ids_to_clear[5000:]
+
+
+def merge_streams(realm: Realm, stream_to_keep: Stream, stream_to_destroy: Stream) -> List[int]:
```
The return type of this isn't a `List` -- that is, it's not variable length, and it's only happenstance that they're all the same type.  We know this is always going to be exactly three elements, so it should be a `Tuple[int, int, int]`.

asah replied:
DONE - thx! for your patience, I'm obviously new to typed python...

alexmv commented on zerver/actions/streams.py:
```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
+        if not users_already_subscribed.get(sub.user_profile_id, False)
+    ]
+
+    # follow a safe order of operations, see
+    # https://github.com/zulip/zulip/pull/22262#pullrequestreview-1017884964
```
If we have to have a comment here arguing for why this order is right (and I'm not convinced we do), then we should inline that argument -- we shouldn't make people fire up their browser to understand why.

asah replied:
DONE - yeah, upon rereading this, I agree it doesn't need an explanation.

alexmv commented on zerver/actions/streams.py:
```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
+    # follow a safe order of operations, see
+    # https://github.com/zulip/zulip/pull/22262#pullrequestreview-1017884964
+
+    # step 1. Subscribe people to the new stream
```
In general, I find these `step n: ...` comments don't add a lot of value -- we're calling one well-named function per step, which makes what's happening pretty clear without additional notes.

asah replied:
DONE - yeah, I agree, makes sense.

alexmv commented on zerver/actions/streams.py:
```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
+        )
+
+    # step 4. Deactivate the stream
+    # note: this might fail in the case that the stream is used by the system, e.g. for
+    # signups, notifications, etc. In this case, in the admin settings either disable these
+    # features or change the stream to something else.
```
This note is useful as a comment, but I might go further than "might fail" and say "might raise a JsonableError" so that it's explicit the _way_ in which it will fail.

But this also confuses me, since below you have a test which shows that merging the notifications stream into the Denmark stream (destroying the notifications stream) is successful, with no exception.  Which is it -- does that raise an exception, or does that unset the notifications stream?

asah replied:
DONE - removed the confusing comment. Sorry, this is a pre-commit before the report-message PR #21821, where it does throw an exception for attempting to archive the reporting stream ([link](https://github.com/zulip/zulip/pull/21821/files#diff-cd6028ed9056ad3e83e6a37f21dfaf1311e45266dc3dc38b710f5f01ab5ec65fR450))

Once report-message is committed, I could see taking on a little refactoring to elevate notification-type-streams to a known type of object with certain properties (e.g. cannot be archived)...

alexmv commented on zerver/actions/streams.py:
```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
+    # signups, notifications, etc. In this case, in the admin settings either disable these
+    # features or change the stream to something else.
+    do_deactivate_stream(stream_to_destroy, acting_user=None)
+    return [len(users_to_activate), count, len(subs_to_deactivate)]
```
As noted above:
```suggestion
    return (len(users_to_activate), count, len(subs_to_deactivate))
```

asah replied:
DONE - thx, this is much cleaner.

alexmv reviewed (commented):
I like splitting this out into a function in `zerver.actions.streams`, especially as it makes it easy to write a test -- but moving _and_ changing the function in the same commit makes it harder to tell what changed.

Can you extract the function in one commit, and then do the reordering in the other?  The one commit message you currently have also doesn't quite cover all of the things that are happening anymore, so splitting it will give you the opportunity to clarify that.

alexmv commented:
> Can you extract the function in one commit, and then do the reordering in the other? The one commit message you currently have also doesn't quite cover all of the things that are happening anymore, so splitting it will give you the opportunity to clarify that.

It looks like this is still one commit with a slightly-out-of-date commit message (this isn't just a "refactor").  Can you split this change into two commits: a refactor and a reordering?

asah commented:
@alexmv split commits - LMK if this helps, thx.

timabbott commented:
Merged, after the following:

* I squashed the lint fixups into the commit that introduced the linter errors (using `git rebase -i upstream/main` -- please do this in the future when fixing a bug that had been introduced in a previous commit in the PR).
* I wrote proper commit messages for the two surviving commits, explaining the motivation for the changes.
* I changed the order of steps to what was suggested in https://github.com/zulip/zulip/pull/22262#pullrequestreview-1017884964, since I agree with the reasoning there.
* I added comments in the reordering commit to explain the motivation for the somewhat unusual order of operations.

Thanks @asah and @alexmv!
